File: apps/api/app/services/content/editorial_publishing_utils.py
"""Build and validate editorial publishing payloads for Shopify."""

from __future__ import annotations

import re
from typing import Any

from app.schemas.content_seo_editorial import (
    EditorialArticlePayload,
    EditorialPublishingPayload,
    EditorialPublishMode,
)

_GID_NUMERIC_RE = re.compile(r"/(\d+)$")


def shopify_gid_numeric_id(gid: str | None) -> str | None:
    if not gid:
        return None
    match = _GID_NUMERIC_RE.search(gid.strip())
    return match.group(1) if match else None


def _coerce_tags(value: Any) -> list[str]:
    if value is None:
        return []
    if isinstance(value, list):
        return [str(v).strip() for v in value if str(v).strip()]
    if isinstance(value, str):
        return [part.strip() for part in value.split(",") if part.strip()]
    return []


def normalize_publishing_payload(raw: dict[str, Any]) -> EditorialPublishingPayload:
    data = dict(raw)
    alias_map = {
        "bodyHtml": "body_html",
        "seoTitle": "seo_title",
        "metaDescription": "meta_description",
        "blogId": "blog_id",
        "blogGid": "blog_gid",
        "imageUrl": "image_url",
        "imageAlt": "image_alt",
        "isPublished": "is_published",
        "publishDate": "publish_date",
        "templateSuffix": "template_suffix",
    }
    for alias, field in alias_map.items():
        if alias in data and field not in data:
            data[field] = data.pop(alias)
    if "tags" in data:
        data["tags"] = _coerce_tags(data["tags"])
    else:
        data.setdefault("tags", [])
    for field in (
        "title",
        "handle",
        "body_html",
        "excerpt",
        "seo_title",
        "meta_description",
        "author",
    ):
        if field in data and data[field] is not None:
            data[field] = str(data[field])
        else:
            data.setdefault(field, "")
    mode = data.get("mode", "draft")
    if mode not in ("draft", "publish_now", "schedule"):
        data["mode"] = "draft"
    data.setdefault("is_published", False)
    return EditorialPublishingPayload.model_validate(data)


def build_publishing_payload_from_article(
    article: EditorialArticlePayload | dict[str, Any],
    *,
    default_blog_id: str | None = None,
    default_blog_gid: str | None = None,
) -> EditorialPublishingPayload:
    if isinstance(article, dict):
        article = EditorialArticlePayload.model_validate(
            {
                "title": article.get("title", ""),
                "handle": article.get("handle", ""),
                "excerpt": article.get("excerpt", ""),
                "body_html": article.get("bodyHtml") or article.get("body_html", ""),
                "body_markdown": article.get("bodyMarkdown") or article.get("body_markdown", ""),
                "seo_title": article.get("seoTitle") or article.get("seo_title", ""),
                "meta_description": article.get("metaDescription")
                or article.get("meta_description", ""),
                "tags": article.get("tags") or [],
                "author_name": article.get("authorName") or article.get("author_name", ""),
            }
        )
    author = (article.author_name or "").strip()
    return EditorialPublishingPayload(
        title=article.title.strip(),
        handle=article.handle.strip(),
        body_html=article.body_html.strip(),
        excerpt=article.excerpt.strip(),
        seo_title=article.seo_title.strip() or article.title.strip(),
        meta_description=article.meta_description.strip(),
        author=author,
        blog_id=default_blog_id,
        blog_gid=default_blog_gid,
        tags=list(article.tags or []),
        mode="draft",
        is_published=False,
        publish_date=None,
    )


def merge_article_into_publishing(
    existing: EditorialPublishingPayload | dict[str, Any],
    article: EditorialArticlePayload | dict[str, Any],
    *,
    overwrite: bool = False,
) -> EditorialPublishingPayload:
    base = (
        normalize_publishing_payload(existing)
        if isinstance(existing, dict)
        else existing
    )
    built = build_publishing_payload_from_article(article)
    if overwrite:
        return built.model_copy(
            update={
                "blog_id": base.blog_id,
                "blog_gid": base.blog_gid,
                "mode": base.mode,
                "is_published": base.is_published,
                "publish_date": base.publish_date,
                "image_url": base.image_url,
                "image_alt": base.image_alt,
                "template_suffix": base.template_suffix,
            }
        )
    updates: dict[str, Any] = {}
    for field in (
        "title",
        "handle",
        "body_html",
        "excerpt",
        "seo_title",
        "meta_description",
        "author",
        "tags",
    ):
        current = getattr(base, field)
        incoming = getattr(built, field)
        if not current and incoming:
            updates[field] = incoming
    if not base.author and built.author:
        updates["author"] = built.author
    return base.model_copy(update=updates)


def validate_publishing_payload(
    payload: EditorialPublishingPayload | dict[str, Any],
    *,
    for_publish: bool = False,
) -> list[str]:
    normalized = (
        normalize_publishing_payload(payload)
        if isinstance(payload, dict)
        else payload
    )
    errors: list[str] = []
    if not normalized.title.strip():
        errors.append("Il titolo è obbligatorio.")
    if not normalized.body_html.strip():
        errors.append("Il contenuto HTML è obbligatorio.")
    if for_publish and not normalized.blog_id and not normalized.blog_gid:
        errors.append("Seleziona un blog Shopify prima di pubblicare.")
    if for_publish and normalized.mode == "schedule":
        errors.append("La pubblicazione programmata non è ancora disponibile.")
    return errors


def build_article_create_input(
    payload: EditorialPublishingPayload,
    *,
    blog_gid: str,
    mode: EditorialPublishMode,
) -> dict[str, Any]:
    from datetime import datetime, timezone

    article_input: dict[str, Any] = {
        "blogId": blog_gid,
        "title": payload.title.strip(),
        "body": payload.body_html.strip(),
    }
    if payload.handle.strip():
        article_input["handle"] = payload.handle.strip()
    if payload.excerpt.strip():
        article_input["summary"] = payload.excerpt.strip()
    if payload.tags:
        article_input["tags"] = payload.tags
    if payload.author.strip():
        article_input["author"] = {"name": payload.author.strip()}
    if payload.template_suffix and payload.template_suffix.strip():
        article_input["templateSuffix"] = payload.template_suffix.strip()
    if payload.image_url and payload.image_url.strip():
        image: dict[str, str] = {"url": payload.image_url.strip()}
        if payload.image_alt and payload.image_alt.strip():
            image["altText"] = payload.image_alt.strip()
        article_input["image"] = image

    if mode == "publish_now":
        article_input["isPublished"] = True
        article_input["publishDate"] = datetime.now(timezone.utc).isoformat()
    else:
        article_input["isPublished"] = False

    return article_input

File: apps/web/src/components/content/editorial/editorial-publishing-utils.ts
import type {
  EditorialArticlePayload,
  EditorialPublishingPayload,
  EditorialPublishMode,
} from "@gcr/shared";

export const DEFAULT_AUTHOR_FALLBACK = "Redazione Solmielato";
export const SEO_TITLE_MAX = 60;
export const META_DESCRIPTION_MAX = 160;
export const SEO_REQUIRED_MESSAGE =
  "SEO title e meta description sono obbligatori per creare un articolo Shopify completo.";

export function emptyEditorialPublishingPayload(): EditorialPublishingPayload {
  return {
    title: "",
    handle: "",
    bodyHtml: "",
    excerpt: "",
    seoTitle: "",
    metaDescription: "",
    author: "",
    blogId: null,
    blogGid: null,
    tags: [],
    imageUrl: null,
    imageAlt: null,
    mode: "draft",
    isPublished: false,
    publishDate: null,
    templateSuffix: null,
    sourceArticleHash: null,
    sourceArticleUpdatedAt: null,
    syncedFromArticleAt: null,
    shopifySeoSynced: null,
    shopifySeoSyncedAt: null,
    shopifySeoError: null,
  };
}

function coerceTags(value: unknown): string[] {
  if (!value) return [];
  if (Array.isArray(value)) {
    return value.map((v) => String(v).trim()).filter(Boolean);
  }
  if (typeof value === "string") {
    return value
      .split(",")
      .map((part) => part.trim())
      .filter(Boolean);
  }
  return [];
}

export function resolveDefaultAuthor(options?: {
  articleAuthorName?: string | null;
  savedAuthor?: string | null;
  shopName?: string | null;
  brandName?: string | null;
}): string {
  const candidates = [
    options?.articleAuthorName,
    options?.savedAuthor,
    options?.shopName,
    options?.brandName,
    options?.brandName ? `Redazione ${options.brandName}` : null,
    DEFAULT_AUTHOR_FALLBACK,
  ];
  for (const candidate of candidates) {
    if (candidate && String(candidate).trim()) {
      return String(candidate).trim();
    }
  }
  return DEFAULT_AUTHOR_FALLBACK;
}

export function parseEditorialPublishingPayload(
  raw: Record<string, unknown> | EditorialPublishingPayload | null | undefined,
): EditorialPublishingPayload {
  if (!raw || Object.keys(raw).length === 0) {
    return emptyEditorialPublishingPayload();
  }
  const mode = raw.mode;
  const normalizedMode: EditorialPublishMode =
    mode === "publish_now" || mode === "schedule" ? mode : "draft";
  return {
    title: String(raw.title ?? ""),
    handle: String(raw.handle ?? ""),
    bodyHtml: String(raw.bodyHtml ?? ""),
    excerpt: String(raw.excerpt ?? ""),
    seoTitle: String(raw.seoTitle ?? ""),
    metaDescription: String(raw.metaDescription ?? ""),
    author: String(raw.author ?? ""),
    blogId: raw.blogId ? String(raw.blogId) : null,
    blogGid: raw.blogGid ? String(raw.blogGid) : null,
    tags: coerceTags(raw.tags),
    imageUrl: raw.imageUrl ? String(raw.imageUrl) : null,
    imageAlt: raw.imageAlt ? String(raw.imageAlt) : null,
    mode: normalizedMode,
    isPublished: Boolean(raw.isPublished),
    publishDate: raw.publishDate ? String(raw.publishDate) : null,
    templateSuffix: raw.templateSuffix ? String(raw.templateSuffix) : null,
    sourceArticleHash: raw.sourceArticleHash ? String(raw.sourceArticleHash) : null,
    sourceArticleUpdatedAt: raw.sourceArticleUpdatedAt
      ? String(raw.sourceArticleUpdatedAt)
      : null,
    syncedFromArticleAt: raw.syncedFromArticleAt ? String(raw.syncedFromArticleAt) : null,
    shopifySeoSynced:
      raw.shopifySeoSynced === undefined || raw.shopifySeoSynced === null
        ? null
        : Boolean(raw.shopifySeoSynced),
    shopifySeoSyncedAt: raw.shopifySeoSyncedAt ? String(raw.shopifySeoSyncedAt) : null,
    shopifySeoError: raw.shopifySeoError ? String(raw.shopifySeoError) : null,
  };
}

export function buildArticleHashCanonical(article: EditorialArticlePayload): string {
  const tags = [...(article.tags ?? [])]
    .map((t) => String(t).trim())
    .filter(Boolean)
    .sort()
    .join(",");
  const payload = {
    title: article.title.trim(),
    handle: article.handle.trim(),
    bodyHtml: article.bodyHtml.trim(),
    excerpt: article.excerpt.trim(),
    seoTitle: article.seoTitle.trim(),
    metaDescription: article.metaDescription.trim(),
    tags,
    authorName: (article.authorName ?? "").trim(),
  };
  return JSON.stringify(payload);
}

export function isPublishingStale(
  article: EditorialArticlePayload | null | undefined,
  publishing: EditorialPublishingPayload | null | undefined,
): boolean {
  if (!article || !publishing) return false;
  const sourceHash = publishing.sourceArticleHash?.trim();
  const articleHash = article.articleHash?.trim();
  if (!sourceHash || !articleHash) return true;
  return sourceHash !== articleHash;
}

/** @deprecated Use isPublishingStale — legacy payloads without hash are now treated as stale. */
export function isPublishingSyncUnknown(
  article: EditorialArticlePayload | null | undefined,
  publishing: EditorialPublishingPayload | null | undefined,
  hasSavedPublishing: boolean,
): boolean {
  if (!article || !publishing || !hasSavedPublishing) return false;
  return isPublishingStale(article, publishing);
}

export function buildPublishingPayloadFromArticle(
  article: EditorialArticlePayload,
  options?: {
    blogId?: string | null;
    blogGid?: string | null;
    shopName?: string | null;
    brandName?: string | null;
    savedAuthor?: string | null;
  },
): EditorialPublishingPayload {
  const author = resolveDefaultAuthor({
    articleAuthorName: article.authorName,
    savedAuthor: options?.savedAuthor,
    shopName: options?.shopName,
    brandName: options?.brandName,
  });
  return {
    title: article.title.trim(),
    handle: article.handle.trim(),
    bodyHtml: article.bodyHtml.trim(),
    excerpt: article.excerpt.trim(),
    seoTitle: article.seoTitle.trim() || article.title.trim(),
    metaDescription: article.metaDescription.trim(),
    author,
    blogId: options?.blogId ?? null,
    blogGid: options?.blogGid ?? null,
    tags: [...(article.tags ?? [])],
    imageUrl: null,
    imageAlt: null,
    mode: "draft",
    isPublished: false,
    publishDate: null,
    templateSuffix: null,
  };
}

export function getPublishingSeoWarnings(payload: EditorialPublishingPayload): string[] {
  const warnings: string[] = [];
  const seoTitle = payload.seoTitle.trim();
  const metaDescription = payload.metaDescription.trim();
  if (seoTitle.length > SEO_TITLE_MAX) {
    warnings.push(
      `SEO title lungo (${seoTitle.length} caratteri; consigliati max ${SEO_TITLE_MAX}).`,
    );
  }
  if (metaDescription.length > META_DESCRIPTION_MAX) {
    warnings.push(
      `Meta description lunga (${metaDescription.length} caratteri; consigliati max ${META_DESCRIPTION_MAX}).`,
    );
  }
  return warnings;
}

export function isPublishingSeoComplete(payload: EditorialPublishingPayload | null | undefined): boolean {
  if (!payload) return false;
  return Boolean(payload.seoTitle.trim() && payload.metaDescription.trim() && payload.handle.trim());
}

export function validatePublishingPayload(
  payload: EditorialPublishingPayload,
  options?: { forPublish?: boolean },
): string[] {
  const errors: string[] = [];
  if (!payload.title.trim()) errors.push("Il titolo è obbligatorio.");
  if (!payload.bodyHtml.trim()) errors.push("Il contenuto HTML è obbligatorio.");
  if (options?.forPublish && !payload.blogId && !payload.blogGid) {
    errors.push("Seleziona un blog Shopify prima di pubblicare.");
  }
  if (options?.forPublish && !payload.author.trim()) {
    errors.push("Autore obbligatorio per creare l'articolo su Shopify.");
  }
  if (options?.forPublish) {
    if (!payload.seoTitle.trim() || !payload.metaDescription.trim()) {
      errors.push(SEO_REQUIRED_MESSAGE);
    }
    if (!payload.handle.trim()) {
      errors.push("Handle obbligatorio per pubblicare su Shopify.");
    }
  }
  return errors;
}

export function validatePublishingPayloadWithWarnings(
  payload: EditorialPublishingPayload,
  options?: { forPublish?: boolean },
): { errors: string[]; warnings: string[] } {
  return {
    errors: validatePublishingPayload(payload, options),
    warnings: options?.forPublish ? getPublishingSeoWarnings(payload) : [],
  };
}

export function tagsToInput(tags: string[]): string {
  return tags.join(", ");
}

export function inputToTags(value: string): string[] {
  return value
    .split(",")
    .map((part) => part.trim())
    .filter(Boolean);
}

export function getPublishStatusLabel(status?: string | null): string {
  switch (status) {
    case "draft_created":
      return "Bozza su Shopify";
    case "published":
      return "Pubblicato";
    case "publish_error":
      return "Errore";
    case "scheduled":
      return "Programmato";
    default:
      return "Non pubblicato";
  }
}

export function extractApiErrorMessage(error: unknown): string | null {
  if (!(error instanceof Error)) return null;
  const detail = error.message.trim();
  if (!detail || detail === "Failed to fetch") return null;
  return detail;
}

export function formatPublishingError(baseMessage: string, error: unknown): string {
  const detail = extractApiErrorMessage(error);
  if (!detail) return baseMessage;
  let message = `${baseMessage} ${detail}`;
  if (message.includes("Field 'seo'") || message.includes("Field `seo`")) {
    message +=
      " Il publisher sta ancora usando un campo GraphQL non supportato. Controllare query Shopify.";
  }
  return message;
}

export function parseStructuredPublishErrorDetail(
  detail: unknown,
): { message: string; code?: string } | null {
  if (!detail || typeof detail !== "object" || Array.isArray(detail)) return null;
  const record = detail as Record<string, unknown>;
  if (typeof record.message === "string") {
    return {
      message: record.message,
      code: typeof record.code === "string" ? record.code : undefined,
    };
  }
  return null;
}
